app/openai_ops.py

```diff
@@ -359,76 +360,82 @@ def context_length(
         raise NotImplementedError(error)
 
 
-# Adapted from https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
+def get_encoding_for_model(model: str):
+    try:
+        return tiktoken.encoding_for_model(model)
+    except KeyError:
+        return tiktoken.get_encoding("cl100k_base")
+
+
+def get_tokens_per_message(model: str):
+    return MODEL_SPECIFIC_TOKENS.get(model, None)
+
+
+def handle_fallback_models(model: str, messages: List[Dict[str, Union[str, Dict[str, str]]]]):
+    model_fallbacks = {
+        GPT_3_5_TURBO_MODEL: GPT_3_5_TURBO_0125_MODEL,
+        GPT_3_5_TURBO_16K_MODEL: GPT_3_5_TURBO_16K_0613_MODEL,
+        GPT_4_MODEL: GPT_4_0613_MODEL,
+        GPT_4_TURBO_MODEL: GPT_4_TURBO_2024_04_09_MODEL,
+        GPT_4_32K_MODEL: GPT_4_32K_0613_MODEL,
+        GPT_4O_MODEL: GPT_4O_2024_05_13_MODEL,
+    }
+    if model in model_fallbacks:
+        return calculate_num_tokens(messages, model=model_fallbacks[model])
+    return None
+
+
+def raise_not_supported_error(model: str):
```

Comment:
this change is not bad but please avoid introducing non-essential changes to your goal. we can do this separately. keep your diff as small as possible